<a href="https://colab.research.google.com/github/enesergen/ML-Fundamentals/blob/main/ML_Day3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Supervised (Tahmin)
"""
Classification:
Regression:
"""
#Unsupervied(Sınıflandırma Clustering)
"""
K-Means
"""


'\nK-Means\n'

In [ ]:
#import pandas lib
import pandas as pd
#data csv url
url="https://raw.githubusercontent.com/tansutosun/MachineLearning/main/cs_accepts.csv"
#read dataset from github csv
df=pd.read_csv(url)
df

TITLE  CHILDREN  PERS_H  AGE  TMADD  TMJOB1  TEL  NMBLOAN  FINLOAN  \
0        R         0       2   46   15.0    33.0  2.0      0.0      0.0   
1        H         4       6   34  144.0    54.0  2.0      1.0      1.0   
2        H         3       5   31  108.0   120.0  2.0      1.0      1.0   
3        R         0       1   39  192.0     6.0  1.0      0.0      0.0   
4        H         3       5   32   48.0   108.0  2.0      2.0      1.0   
...    ...       ...     ...  ...    ...     ...  ...      ...      ...   
2995     H         0       1   26  192.0     6.0  2.0      2.0      1.0   
2996     H         0       1   34   12.0    30.0  2.0      2.0      0.0   
2997     H         0       1   32    9.0    24.0  2.0      2.0      1.0   
2998     R         0       1   31    3.0    12.0  2.0      2.0      1.0   
2999     R         0       2   27   36.0    36.0  1.0      2.0      1.0   

      INCOME  EC_CARD STATUS  BUREAU  LOCATION  LOANS  REGN  DIV    CASH  \
0        0.0      1.0      V     3.0       1.0      0     0  0.0  2000.0   
1     3200.0      0.0      V     1.0       1.0      2     4  1.0  6000.0   
2     3300.0      0.0      V     1.0       1.0      3     4  1.0     0.0   
3     1500.0      0.0      W     1.0       1.0      1     0  0.0  2500.0   
4        0.0      1.0      V     1.0       1.0      1     0  0.0  2500.0   
...      ...      ...    ...     ...       ...    ...   ...  ...     ...   
2995  2300.0      0.0      U     1.0       1.0      2     2  1.0  5000.0   
2996  2500.0      0.0      U     1.0       1.0      1     5  1.0  3000.0   
2997  2800.0      0.0      G     1.0       1.0      3     5  1.0  2500.0   
2998  2100.0      0.0      G     1.0       1.0      1     4  1.0  4000.0   
2999  2200.0      0.0      V     1.0       1.0      1     4  1.0  2500.0   

               PRODUCT  RESID      NAT              PROF              CAR  \
0      Radio, TV, Hifi  Lease   German            Others              Car   
1                  NaN  Owner  Turkish               NaN              Car   
2                  NaN  Lease  Turkish            Others              Car   
3     Furniture,Carpet  Lease   German            Others  Without Vehicle   
4     Furniture,Carpet  Lease   German  Civil Service, M              Car   
...                ...    ...      ...               ...              ...   
2995   Radio, TV, Hifi  Lease   German            Others  Without Vehicle   
2996  Furniture,Carpet  Lease   German            Others              Car   
2997   Radio, TV, Hifi  Lease   German            Others              Car   
2998  Furniture,Carpet  Lease   German            Others              Car   
2999   Radio, TV, Hifi  Lease   German  State,Steel Ind,  Without Vehicle   

                CARDS  GB  _freq_  
0         Cheque card   0      30  
1     no credit cards   1       1  
2     no credit cards   1       1  
3     no credit cards   1       1  
4         Cheque card   0      30  
...               ...  ..     ...  
2995  no credit cards   1       1  
2996  no credit cards   0      30  
2997  no credit cards   1       1  
2998  no credit cards   0      30  
2999  no credit cards   1       1  

[3000 rows x 26 columns]

In [ ]:
df.isnull().any()

TITLE       False
CHILDREN    False
PERS_H      False
AGE         False
TMADD       False
TMJOB1      False
TEL         False
NMBLOAN     False
FINLOAN     False
INCOME      False
EC_CARD     False
STATUS      False
BUREAU      False
LOCATION    False
LOANS       False
REGN        False
DIV         False
CASH        False
PRODUCT      True
RESID        True
NAT         False
PROF         True
CAR         False
CARDS       False
GB          False
_freq_      False
dtype: bool

In [ ]:
#kendim yaptım
product_dummies=pd.get_dummies(df['PRODUCT'],dummy_na=True)
product_dummies

Cars  Dept. Store,Mail  Furniture,Carpet  Leisure  Others  \
0     False             False             False    False   False   
1     False             False             False    False   False   
2     False             False             False    False   False   
3     False             False              True    False   False   
4     False             False              True    False   False   
...     ...               ...               ...      ...     ...   
2995  False             False             False    False   False   
2996  False             False              True    False   False   
2997  False             False             False    False   False   
2998  False             False              True    False   False   
2999  False             False             False    False   False   

      Radio, TV, Hifi    NaN  
0                True  False  
1               False   True  
2               False   True  
3               False  False  
4               False  False  
...               ...    ...  
2995             True  False  
2996            False  False  
2997             True  False  
2998            False  False  
2999             True  False  

[3000 rows x 7 columns]

In [ ]:
import statsmodels.api as sm
y=df[['INCOME']]
y

INCOME
0        0.0
1     3200.0
2     3300.0
3     1500.0
4        0.0
...      ...
2995  2300.0
2996  2500.0
2997  2800.0
2998  2100.0
2999  2200.0

[3000 rows x 1 columns]

In [ ]:
#girdi olarak kullanmak istediğimiz özellikler
x=df[['TMJOB1','AGE','CASH']]
x

TMJOB1  AGE    CASH
0       33.0   46  2000.0
1       54.0   34  6000.0
2      120.0   31     0.0
3        6.0   39  2500.0
4      108.0   32  2500.0
...      ...  ...     ...
2995     6.0   26  5000.0
2996    30.0   34  3000.0
2997    24.0   32  2500.0
2998    12.0   31  4000.0
2999    36.0   27  2500.0

[3000 rows x 3 columns]

In [ ]:
#sabit constant ekledik kolonlara ek olarak
x=sm.add_constant(x)
x

const  TMJOB1  AGE    CASH
0       1.0    33.0   46  2000.0
1       1.0    54.0   34  6000.0
2       1.0   120.0   31     0.0
3       1.0     6.0   39  2500.0
4       1.0   108.0   32  2500.0
...     ...     ...  ...     ...
2995    1.0     6.0   26  5000.0
2996    1.0    30.0   34  3000.0
2997    1.0    24.0   32  2500.0
2998    1.0    12.0   31  4000.0
2999    1.0    36.0   27  2500.0

[3000 rows x 4 columns]

In [ ]:
#model eğitimi yapıldı
model=sm.OLS(y,x).fit()

In [ ]:
#modelin analiz sonucunu göstermek için yazıldı
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 INCOME   R-squared:                       0.010
Model:                            OLS   Adj. R-squared:                  0.009
Method:                 Least Squares   F-statistic:                     10.47
Date:                Wed, 08 May 2024   Prob (F-statistic):           7.54e-07
Time:                        11:33:16   Log-Likelihood:                -27487.
No. Observations:                3000   AIC:                         5.498e+04
Df Residuals:                    2996   BIC:                         5.501e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1822.0586    139.775     13.036      0.000    1547.994    2096.123
TMJOB1         1.2187      0.364      3.352      0.001       0.506       1.932
AGE            0.2501      4.128      0.061      0.952      -7.843       8.344
CASH           0.0278      0.007      4.192      0.000       0.015       0.041
==============================================================================
Omnibus:                     7412.002   Durbin-Watson:                   1.920
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        144438809.146
Skew:                          25.654   Prob(JB):                         0.00
Kurtosis:                    1076.722   Cond. No.                     2.27e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.27e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""


Yukarıdaki sonuçlara göre age 	0.952 değeri ile etkisiz olduğunu anlıyoruz
çünkü katsayı değer aralığı -7.843,8.344 aralığındadır yani 0 da olabilir bu sayede denkleme etkisi olmaz
bu yüzden bu özelliği kullanmak çokta doğru değildir.



In [ ]:
predicted_df=pd.DataFrame(model.predict(x),columns=['predicted'])
predicted_df

predicted
0     1929.381679
1     2063.178423
2     1976.057011
3     1908.626351
4     2031.185228
...           ...
2995  1974.878258
2996  1950.525829
2997  1928.812825
2998  1955.639796
2999  1942.187119

[3000 rows x 1 columns]

In [ ]:
input_df=pd.DataFrame(x)
input_df

const  TMJOB1  AGE    CASH
0       1.0    33.0   46  2000.0
1       1.0    54.0   34  6000.0
2       1.0   120.0   31     0.0
3       1.0     6.0   39  2500.0
4       1.0   108.0   32  2500.0
...     ...     ...  ...     ...
2995    1.0     6.0   26  5000.0
2996    1.0    30.0   34  3000.0
2997    1.0    24.0   32  2500.0
2998    1.0    12.0   31  4000.0
2999    1.0    36.0   27  2500.0

[3000 rows x 4 columns]

In [ ]:
output_df=pd.DataFrame(y)
output_df

INCOME
0        0.0
1     3200.0
2     3300.0
3     1500.0
4        0.0
...      ...
2995  2300.0
2996  2500.0
2997  2800.0
2998  2100.0
2999  2200.0

[3000 rows x 1 columns]

In [ ]:
final_df=pd.concat([input_df,output_df,predicted_df],axis=1)
final_df

const  TMJOB1  AGE    CASH  INCOME    predicted
0       1.0    33.0   46  2000.0     0.0  1929.381679
1       1.0    54.0   34  6000.0  3200.0  2063.178423
2       1.0   120.0   31     0.0  3300.0  1976.057011
3       1.0     6.0   39  2500.0  1500.0  1908.626351
4       1.0   108.0   32  2500.0     0.0  2031.185228
...     ...     ...  ...     ...     ...          ...
2995    1.0     6.0   26  5000.0  2300.0  1974.878258
2996    1.0    30.0   34  3000.0  2500.0  1950.525829
2997    1.0    24.0   32  2500.0  2800.0  1928.812825
2998    1.0    12.0   31  4000.0  2100.0  1955.639796
2999    1.0    36.0   27  2500.0  2200.0  1942.187119

[3000 rows x 6 columns]

In [ ]:
final_df['error']=final_df['predicted']-final_df['INCOME']
final_df

const  TMJOB1  AGE    CASH  INCOME    predicted        error
0       1.0    33.0   46  2000.0     0.0  1929.381679  1929.381679
1       1.0    54.0   34  6000.0  3200.0  2063.178423 -1136.821577
2       1.0   120.0   31     0.0  3300.0  1976.057011 -1323.942989
3       1.0     6.0   39  2500.0  1500.0  1908.626351   408.626351
4       1.0   108.0   32  2500.0     0.0  2031.185228  2031.185228
...     ...     ...  ...     ...     ...          ...          ...
2995    1.0     6.0   26  5000.0  2300.0  1974.878258  -325.121742
2996    1.0    30.0   34  3000.0  2500.0  1950.525829  -549.474171
2997    1.0    24.0   32  2500.0  2800.0  1928.812825  -871.187175
2998    1.0    12.0   31  4000.0  2100.0  1955.639796  -144.360204
2999    1.0    36.0   27  2500.0  2200.0  1942.187119  -257.812881

[3000 rows x 7 columns]

In [ ]:
#Logistic Regression

In [ ]:
Data=df[['INCOME','TMJOB1','CASH','GB']]
Data

INCOME  TMJOB1    CASH  GB
0        0.0    33.0  2000.0   0
1     3200.0    54.0  6000.0   1
2     3300.0   120.0     0.0   1
3     1500.0     6.0  2500.0   1
4        0.0   108.0  2500.0   0
...      ...     ...     ...  ..
2995  2300.0     6.0  5000.0   1
2996  2500.0    30.0  3000.0   0
2997  2800.0    24.0  2500.0   1
2998  2100.0    12.0  4000.0   0
2999  2200.0    36.0  2500.0   1

[3000 rows x 4 columns]

In [ ]:
Y=Data[['GB']]
Y

GB
0      0
1      1
2      1
3      1
4      0
...   ..
2995   1
2996   0
2997   1
2998   0
2999   1

[3000 rows x 1 columns]

In [ ]:
X=Data[['INCOME','TMJOB1','CASH']]
X

INCOME  TMJOB1    CASH
0        0.0    33.0  2000.0
1     3200.0    54.0  6000.0
2     3300.0   120.0     0.0
3     1500.0     6.0  2500.0
4        0.0   108.0  2500.0
...      ...     ...     ...
2995  2300.0     6.0  5000.0
2996  2500.0    30.0  3000.0
2997  2800.0    24.0  2500.0
2998  2100.0    12.0  4000.0
2999  2200.0    36.0  2500.0

[3000 rows x 3 columns]

In [ ]:
from sklearn.linear_model import LogisticRegression
model=LogisticRegression(random_state=42).fit(X,Y)


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
model.predict(X)

array([0, 1, 1, ..., 1, 1, 1])

In [ ]:
#Burada array listesinin ilk elemanı [0.53204014, 0.46795986] ilk değer 0 olma ihtimali diğeri 1 olma ihtimali
model.predict_proba(X)

array([[0.53204014, 0.46795986],
       [0.45127515, 0.54872485],
       [0.49537853, 0.50462147],
       ...,
       [0.43280955, 0.56719045],
       [0.44778916, 0.55221084],
       [0.4627866 , 0.5372134 ]])

In [ ]:
y_ped_prob=model.predict_proba(X)[:,1]
y_ped_prob
